From 1093128c646b154a14d89321454f5361c0e616b4 Mon Sep 17 00:00:00 2001
Message-Id: <1093128c646b154a14d89321454f5361c0e616b4.1600854717.git.liu.denton@gmail.com>
From: Denton Liu <liu.denton@gmail.com>
Date: Tue, 22 Sep 2020 15:58:38 -0700
Subject: [PATCH] pull: do not warn when opt_ff is explicitly specified
To: Git Mailing List <git@vger.kernel.org>
Cc: Alex Henrie <alexhenrie24@gmail.com>

In d18c950a69 (pull: warn if the user didn't say whether to rebase or to
merge, 2020-03-09), `git pull` was taught to warn users if they
have `pull.rebase` unset or `pull.ff != "only"`. However, this warning
is a little too eager about happening.

If the warning is silenced by specifying `pull.ff = "only"`, as
instructed, the warning will arise again if the user runs
something like `git pull --no-ff`. However, the warning should not
happen as the user clearly knows what they're doing.

Don't display the warning if opt_ff is explicitly set by a command-line
option given by a user.

Signed-off-by: Denton Liu <liu.denton@gmail.com>
---
 builtin/pull.c | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/builtin/pull.c b/builtin/pull.c
index 015f6ded0b..307b4b5d21 100644
--- a/builtin/pull.c
+++ b/builtin/pull.c
@@ -83,6 +83,7 @@ static char *opt_commit;
 static char *opt_edit;
 static char *cleanup_arg;
 static char *opt_ff;
+static int opt_ff_explicit;
 static char *opt_verify_signatures;
 static int opt_autostash = -1;
 static int config_autostash;
@@ -344,7 +345,7 @@ static enum rebase_type config_get_rebase(void)
 	if (!git_config_get_value("pull.rebase", &value))
 		return parse_config_rebase("pull.rebase", value, 1);
 
-	if (opt_verbosity >= 0 &&
+	if (!opt_ff_explicit && opt_verbosity >= 0 &&
 	    (!opt_ff || strcmp(opt_ff, "--ff-only"))) {
 		warning(_("Pulling without specifying how to reconcile divergent branches is\n"
 			"discouraged. You can squelch this message by running one of the following\n"
@@ -933,6 +934,8 @@ int cmd_pull(int argc, const char **argv, const char *prefix)
 
 	if (!opt_ff)
 		opt_ff = xstrdup_or_null(config_get_ff());
+	else
+		opt_ff_explicit = 1;
 
 	if (opt_rebase < 0)
 		opt_rebase = config_get_rebase();
-- 
2.28.0.760.g8d73e04208

